File: Release/WetiMatlabFunctions/DataFunctions/CalculateLifeTimeWeightedMean.m
function ProcessResults = CalculateLifeTimeWeightedMean(ProcessResults,Statistics,WindSpeedBins,InputID,FilterID,OutputID,OutputID_PerBin,options)
% Calculates life-time weighted DEL.
% Function for CalculateProcessResults.

% inputs
arguments
    ProcessResults              struct
    Statistics                  table
    WindSpeedBins               (1,:) double
    InputID                     char
    FilterID                    char            = 'WindSpeedsBinsFilter' %   default from DataFilterWindSpeedBins, but could be different for real data
    OutputID                    char            = ['LTW_',InputID]
    OutputID_PerBin             char            = ['LTW_',InputID,'_PerBin']
    options.A_Weibull           (1,1) double    = 2/sqrt(pi)*10;    % Class I
    options.k_Weibull           (1,1) double    = 2                 % Rayleigh distribution                
end
% TODO: 
% * options.Distribution to feed in site-specific wind distribution
% * options.WindClass char {mustBeMember(options.WindClass,{'I','II','III'})} 

% get dimensions
nWindSpeedBins              = size(WindSpeedBins,2);
Filter                      = ProcessResults.(FilterID);
[~,nFilters]                = size(Filter);

% local variables for options
A                           = options.A_Weibull;
k                           = options.k_Weibull;

% check dimension of filter
if nFilters~=nWindSpeedBins
    error(['Dimension of Filter ',FilterID,' does not fit to WindSpeedBins input.']);
end

% loop over bins
MEAN        = Statistics.(InputID);
MEAN_PerBin = NaN(1,nWindSpeedBins);
for iWindSpeedBin = 1:nWindSpeedBins
    MEAN_PerBin(iWindSpeedBin) = mean(MEAN(Filter(:,iWindSpeedBin)));    
end

% life time weighting MEAN
Distribution        = k/A*(WindSpeedBins/A).^(k-1).*exp(-(WindSpeedBins/A).^k);
Weights             = Distribution./sum(Distribution);
LTW_MEAN_PerBin     = MEAN_PerBin.*Weights;
LTW_MEAN            = sum(LTW_MEAN_PerBin);

% store in ProcessResults
ProcessResults.(OutputID)           = LTW_MEAN;
ProcessResults.(OutputID_PerBin)    = LTW_MEAN_PerBin;

end

File: Release/PythonFunctions/CalculateLifeTimeWeightedMean.py
import numpy as np

def CalculateLifeTimeWeightedMean(process_results, statistics, wind_speed_bins, input_id, filter_id='WindSpeedsBinsFilter', output_id=None, output_id_per_bin=None, options=None):
    # Function for CalculateProcessResults.

    # inputs
    if options is None:
        options = {}
    if output_id is None:
        output_id = 'LTW_' + input_id
    if output_id_per_bin is None:
        output_id_per_bin = 'LTW_' + input_id + '_PerBin'

    # local variables for options
    A = options.get('A_Weibull', 2 / np.sqrt(np.pi) * 10)
    k = options.get('k_Weibull', 2)

    # get dimensions
    n_wind_speed_bins = len(wind_speed_bins)
    filter_data = process_results[filter_id]
    _, n_filters = filter_data.shape

    # check dimension of filter
    if n_filters != n_wind_speed_bins:
        raise ValueError(f'Dimension of Filter {filter_id} does not fit WindSpeedBins input.')

    # loop over bins
    mean_data = statistics[input_id]
    mean_per_bin = np.full(n_wind_speed_bins, np.nan)
    for i_wind_speed_bin in range(n_wind_speed_bins):
        filter_this_bin = filter_data[:, i_wind_speed_bin]
        mean_per_bin[i_wind_speed_bin] = np.mean(mean_data[filter_this_bin])

    # life time weighting MEAN
    distribution = k / A * (wind_speed_bins / A)**(k - 1) * np.exp(-(wind_speed_bins / A)**k)
    weights = distribution / np.sum(distribution)
    ltw_mean_per_bin = mean_per_bin * weights
    ltw_mean = np.sum(ltw_mean_per_bin)

    # store in ProcessResults
    process_results[output_id] = ltw_mean
    process_results[output_id_per_bin] = ltw_mean_per_bin

    return process_results
